### PR TITLE
Snippet support

### PR DESCRIPTION
**Pull request by mg979**
Generator can return the annotation lines as an anonymous snippet body,
where placeholders are sequential ($1, $2, ...).

neogen.generate() will return 2 values: the snippet lines and the row
where the annotation should be inserted.

### CODE REVIEW

**Comment by danymat** [comment]
Hello, very clever !!
Do you mind sharing how you integrate with a snippet engine ? Personally, I use luasnip

**Comment by mg979** [comment]
I don't know with luasnip, it requires a snippet engine that supports anonymous snippets. I think vim-vsnip does, personally I use my own fork of [nvim-snippy](https://github.com/mg979/nvim-snippy) and I also support them there. You can see how I use it [here](https://github.com/mg979/dotfiles/blob/master/Cloud/nvim/config/lua/settings/neogen.lua).

**Comment by mg979** [comment]
Peek 23-02-2022 20-38

**Comment by danymat** [comment]
It seems to work with luasnip:

```lua
local ls = require("luasnip")


vim.keymap.set("n", "<Leader>cx", function()
	local snip, row = require("neogen").generate(nil, true)
	if not snip then
		return
	end
	ls.lsp_expand("\n" .. table.concat(snip, "\n"), { pos = { row - 1, 0 } })
end)


-- Test here:

--- 
---@param salut 
local function test(salut) end
```

Maybe @L3MON4D3 you could be interested here !

**Comment by danymat** [comment]
I like it a lot, and it can be improved a lot, leveraging default text for cursor position !
I have some ideas:

- Passing some default text to templates (`{$1:...}`)
- Adding most common snippet engines integration out of the box

But before that we have to make some adjustments in this PR:
- Write some documentation for neogen help file + README
- I prefer using the `opts` table instead, something like this: `require'neogen'.generate({ snippet = true })`
- We have to return the `col` too

File: lua/neogen/generator.lua
Comment (max397574) [inline lua/neogen/generator.lua:215]: couldn't you just use `i` instead of `ph` here?

Comment (max397574) [reply lua/neogen/generator.lua:215]: but you don't check for this

**Comment by mg979** [comment]
>     * Passing some default text to templates (`{$1:...}`)

It would need lsp, getting the parameter name, etc, not sure it's feasible in all circumstances.
 
> But before that we have to make some adjustments in this PR:
> 
>     * Write some documentation for neogen help file + README
> 
>     * I prefer using the `opts` table instead, something like this: `require'neogen'.generate({ snippet_engine = "luasnip" })`

Done.
 
>     * We have to return the `col` too

Where is that? I see `last_col` but I'm not sure it's what is needed.

**Comment by L3MON4D3** [comment]
> It seems to work with luasnip:
> 
> ```lua
> local ls = require("luasnip")
> 
> 
> vim.keymap.set("n", "<Leader>cx", function()
> 	local snip, row = require("neogen").generate(nil, true)
> 	if not snip then
> 		return
> 	end
> 	ls.lsp_expand("\n" .. table.concat(snip, "\n"), { pos = { row - 1, 0 } })
> end)
> 
> 
> -- Test here:
> 
> --- 
> ---@param salut 
> local function test(salut) end
> ```
> 
> Maybe @L3MON4D3 you could be interested here !

Looks like it should work! Also looking forward to using that, just one key for jumping around sounds great :D

**Comment by mg979** [comment]
Anyway, currently the code is checking if there is an empty space at the end of the line and adds the placeholder there, but maybe there can be placeholders in the middle of a line? I think it should use `mark_pos` instead.

File: lua/neogen/generator.lua
Comment (max397574) [reply lua/neogen/generator.lua:215]: @mg979  ^^

**Comment by mg979** [comment]
Built-in integration could be done like this (more or less):

```diff
diff --git a/lua/neogen/init.lua b/lua/neogen/init.lua
index 8d26fa1..80c0c5b 100644
--- a/lua/neogen/init.lua
+++ b/lua/neogen/init.lua
@@ -131,7 +131,19 @@ neogen.generate = function(opts)
         return
     end
 
-    return require("neogen.generator")(vim.bo.filetype, opts and opts.type, opts and opts.snippet)
+    local lines, row = require("neogen.generator")(vim.bo.filetype, opts and opts.type, opts and opts.snippet)
+
+    if opts.snippet == 'luasnip' then
+      local ok, ls = pcall(require, 'luasnip')
+      if ok then
+        ls.lsp_expand("\n" .. table.concat(lines, "\n"), { pos = { row - 1, 0 } })
+      else
+        -- fallback to built-in expansion
+        require("neogen.generator")(vim.bo.filetype, opts and opts.type)
+      end
+    end
+
+    return ret
 end
 
 -- Expose more API  ============================================================
```

**Comment by danymat** [comment]
> Built-in integration could be done like this (more or less):
> 
> ```diff
> diff --git a/lua/neogen/init.lua b/lua/neogen/init.lua
> index 8d26fa1..0865304 100644
> --- a/lua/neogen/init.lua
> +++ b/lua/neogen/init.lua
> @@ -131,7 +131,23 @@ neogen.generate = function(opts)
>          return
>      end
>  
> -    return require("neogen.generator")(vim.bo.filetype, opts and opts.type, opts and opts.snippet)
> +    if not opts or not opts.snippet then
> +      return require("neogen.generator")(vim.bo.filetype, opts and opts.type)
> +    end
> +
> +    local lines, row = require("neogen.generator")(vim.bo.filetype, opts.type, opts.snippet)
> +
> +    if opts.snippet == 'luasnip' then
> +      local ok, ls = pcall(require, 'luasnip')
> +      if ok then
> +        ls.lsp_expand("\n" .. table.concat(lines, "\n"), { pos = { row - 1, 0 } })
> +      else
> +        -- fallback to built-in expansion
> +        require("neogen.generator")(vim.bo.filetype, opts and opts.type)
> +      end
> +    end
> +
> +    return lines
>  end
>  
>  -- Expose more API  ============================================================
> ```

I prefer to change it inside the generator, because inside it, it's still setting marks. I will take a look at it rn with you

**Comment by mg979** [comment]
If you set the `snippet` option it doesn't set marks, the generator returns before adding them. Anyway this example is probably bad.

File: lua/neogen/generator.lua
Comment (mg979) [reply lua/neogen/generator.lua]: There I was adding placeholders only if the line ended with an empty space. I changed it by now to use marks positions anyway.

**Comment by danymat** [comment]
@mg979 it seems like you didn't allow edits by maintainer, can you take a look ?

**Comment by danymat** [comment]
@mg979 I just added support for bundled snippet engines integration, can you take a look ? You can add support for the [nvim-snippy](https://github.com/mg979/nvim-snippy) plugin if you want to try out

**Comment by danymat** [comment]
> The problem is that only my fork of nvim-snippy supports this kind of expansion, so it's not a good candidate. Can't you just return the generated lines if the `snippet_engine` is not among the ones supported out of the box? It seems you're throwing an error currently.

Yes, I'm thinking to reword the `snippet` option to `snippet_engine`, and provide a `return_snippet` option to return the snippet, as well as the position, what do you think ?

**Comment by mg979** [comment]
It would be good.

**Comment by danymat** [comment]
Can you check it out ?

**Comment by mg979** [comment]
It's working ok!

**Comment by mg979** [comment]
The README.md still has the old documentation.

**Comment by danymat** [comment]
@L3MON4D3, it seems that the lsp_expand is having some inconsistencies:

```lua
    local expanded_snip = "\n" .. table.concat(snip, "\n")
    P(expanded_snip)
    luasnip.lsp_expand(expanded_snip, { pos = { pos[1] - 1, pos[2] } })
```

results in:

Result of P: `"\n    --- $1\n    ---@param salut $2"`

Before annotation:

```lua
    local function tes(salut) end
```

After annotation:

```lua
        --- 
        ---@param salut 
    local function tes(salut) end
```

Do you know why it offsets like this ?

**Comment by danymat** [comment]
> The README.md still has the old documentation.

Just fixed

**Comment by mg979** [comment]
> ```lua
>     local expanded_snip = "\n" .. table.concat(snip, "\n")
>     P(expanded_snip)
>     luasnip.lsp_expand(expanded_snip, { pos = { pos[1] - 1, pos[2] } })
> ```

Which indentation level is that? Did it work differently in your first attempt (your 2nd post)?

**Comment by danymat** [comment]
I don't think you should remove indentation for snippet, as it's the same for the content 

In fact, we place the content at col 0 because the indentation is created with spaces inside the content

**Comment by mg979** [comment]
> I don't think you should remove indentation for snippet

I think the first level of indentation should be removed, but the same for all lines (not like I did).
Unless the snippet plugin doesn't handle that on its own. Generally snippet don't have extra indentation, the snippet plugin adds it where necessary, so a snippet template shouldn't have any extra indentation.

But there are other problems if there are multiple placeholders in the same line, I'm seeing that in python, I'm trying to fix that too.

**Comment by danymat** [comment]
> Unless the snippet plugin doesn't handle that on its own.

I think we should wait on @L3MON4D3's response, because if we remove indentation, it'll mess up in creating properly indented comments. You can try offsetting a function in your file, and calling neogen on top of it.

**Comment by mg979** [comment]
I think we just try what works and what not. I don't think there's anything wrong on luasnip's part, it's the snippet that it's given that is malformed.

**Comment by L3MON4D3** [comment]
> 

The spaces are replaced with tabs, right? If so, not sure where they is coming from, it should be expanded with leading spaces

**Comment by L3MON4D3** [comment]
> I think we just try what works and what not. I don't think there's anything wrong on luasnip's part, it's the snippet that it's given that is malformed.

Normally luasnip should handle leading indentation just fine, very weird that it fails here..

**Comment by L3MON4D3** [comment]
I'll look into what's going wrong with our parser, normally the leading indentation should be present in the expanded snippet, at least for luasnip

**Comment by mg979** [comment]
@danymat did you try with last commit? I tried it with luasnip and it's ok for lua (not for python).

**Comment by mg979** [comment]
Peek 24-02-2022 12-32
Peek 24-02-2022 12-34

For me it works with nvim-snippy. But I have to create the new line, and put the cursor at the correct position, like this:
```lua
    local lines, row = require'neogen'.generate({ return_snippet = true })
    if not lines then
      return
    end
    local co = vim.api.nvim_replace_termcodes('<c-o>', true, true, true)
    vim.cmd(tostring(row + 1))
    vim.cmd('normal! O_' .. co .. 'x')
    require 'snippy'.expand_snippet({body = lines, kind = 'snipmate'})
```
 I don't know which are the required steps for luasnip.

**Comment by danymat** [comment]
In python, behavior is inconsistent:

Capture d’écran 2022-02-24 à 14 53 10

Capture d’écran 2022-02-24 à 14 53 00

**Comment by danymat** [comment]
This is what should be done:

- keep the offset, and do not touch the generated output
- correctly place the snippet on provided row, at col 0. (col 0 because the offset is already generated in the output)

**Comment by L3MON4D3** [comment]
Ahhhh, I see what's going on, see [this line](https://github.com/L3MON4D3/LuaSnip/blob/40d9d58938d2a24da141b1da07550a36b144d538/lua/luasnip/nodes/snippet.lua#L329): even if the snippet should expand at a position other than the current cursor, luasnip always indents the snippet using the indent of the current line :sweat_smile: 
I pushed a fix, it should work consistently now.

**Comment by mg979** [comment]
With the latest luasnip fix, this work for me. Note that I'm still stripping the indent, but returning it so that a new line can be created with the correct indent, then the snippet is expanded at that position. Not removing the indent from the template causes issues in nested/indented functions, because the snippet engine thinks that indent is part of the snippet, and indents it again.

**Comment by mg979** [comment]
This is luasnip:

Peek 24-02-2022 16-48
Peek 24-02-2022 16-49

File: lua/neogen/snippet.lua
Comment (danymat) [inline lua/neogen/snippet.lua:54]: You can remove the indent variable and just use pos[2], which is the `col` position

Comment (danymat) [inline lua/neogen/snippet.lua:54]: You can remove the indent variable and just use pos[2], which is the `col` position

Comment (mg979) [reply lua/neogen/snippet.lua]: Done.

File: lua/neogen/generator.lua
Comment (danymat) [inline lua/neogen/generator.lua:223]: I still don't get why you want to remove indented spaces from the generated content, only to add them back afterwards with col. Why just avoid deleting spaces inside content, and append at col 0 ? 

Maybe I'm missing something here

**Comment by mg979** [comment]
I'll quote myself from above.

> Not removing the indent from the template causes issues in nested/indented functions, because the snippet engine thinks that indent is part of the snippet, and indents it again.

I tried avoiding doing that, and it messed up annotations that weren't at 0 indent. Because snippets plugins work like that, they're supposed to receive a non-indented snippet, and they indent it if needed.

The thing that bothers you is maybe that it's doing double work, first adding the indent, then removing it. But this is because the previous code added the indent to the template already, because it wasn't meant to be used in some other way. You can refactor it if you want, adding indent at a later stage. But passing an indented snippet to the snippet plugin will mess up things.

Thing is, like this it works, in other ways it doesn't.

**Comment by mg979** [comment]
Ok you may be right, this should also work.

**Comment by danymat** [comment]
I can confirm it's working on my end with luasnip integration. I like the way this PR is shaping up ! (also, thanks @L3MON4D3 for the quick fix !)

I still want to integrate this feature:

> Passing some default text to templates ({$1:...})

Which is a way of formulating this issue: https://github.com/danymat/neogen/issues/49

**Comment by mg979** [comment]
I think this would add more problems: for example, you have a parameter and for default text you use the parameter type that you get from lsp. But then you either change it (meaning you have to rewrite it), or you jump to next, that is probably another parameter. If you want to add some text (description, or else), then you'd have to add another tabstop that you can jump to, right after the placeholder. It think this PR isn't the right place to do that anyway.

**Comment by danymat** [comment]
You're right, but I think we misunderstood each other. The issue is related to showing some text for when you jump on them, like this: https://github.com/danymat/neogen/issues/34#issuecomment-1024912939

I know that lua snip handle those placeholder texts

**Comment by mg979** [comment]
Ah ok, if they're supposed to be changed in any case it should be doable.

**Comment by danymat** [comment]
Indeed, but I think this feature will require a lot more of rewrite, so I propose to review this PR as it is right now, and merge it if we find it stable enough.

**Comment by mg979** [comment]
Minor change to the documentation since `col` isn't needed anymore (it's always `0`).

I had some small success with the other idea (`[TODO]` default text) but later.

**Comment by danymat** [comment]
> I had some small success with the other idea ([TODO] default text) but later.

Nice, if you're interested we could open a WIP later, anyways this PR is refreshing and it was a very good idea to support snippets, thanks 

By the way, I too did play around with "TODO", and this is my findings:
- It works, and I had the idea of creating a new key in the 3rd argument of a template (like this:)

```lua
    { nil, "# $1", { type = { "class", "func" }, placeholders = { "TODO:description" } } },
  ...
``
- The most difficult will be the parsing. It should be done before generating the content, when the code still has the visibility around the template options (in the `generate_content` call)

**Review by danymat — APPROVED** [review]

**Comment by mg979** [comment]
My idea was that the placeholder text should end up in `mark_pos`, that is, when the placeholder is added, it should contain the information for the default text.

Then when the snippet is built, this would work:
```lua
--- Converts a template to a lsp-compatible snippet
---@param template string[] the generated annotations to parse
---@param marks table generated marks for the annotations
---@param pos table a tuple of row,col
---@return table resulting snippet lines
---@private
snippet.to_snippet = function(template, marks, pos)
    local offset, ph = {}, {}
    for i, m in ipairs(marks) do
        local r, col = m[1] - pos[1] + 1, m[2]
        ph[i] = m.text and string.format('${%d:%s}', i, m.text) or '$' .. i
        if offset[r] then
            offset[r] = offset[r] + ph[i - 1]:len() + 1
        else
            offset[r] = 0
        end
        local pre = template[r]:sub(1, col + offset[r])
        template[r] = pre .. ph[i] .. template[r]:sub(col + 1 + offset[r])
    end
    return template
end
```

The hard part for me was to actually setting `mark.text`, because template and marks are generated in different places.

**Comment by danymat** [comment]
Yeah this would work, though clearly find this way bypassing every safe coding instructions 😆 
If we are to pass more informations to marks, maybe we should create a specific table handling: mark + placeholder + whatever

**Comment by mg979** [comment]
I got it working (the logic can be improved):
Peek 24-02-2022 23-49
Peek 24-02-2022 23-51

**Comment by danymat** [comment]
Very nice, this is exactly the idea. I can help improving the logic

**Comment by danymat** [comment]
Nice ! Feel free to share the code, so we can improve the logic together

**Comment by mg979** [comment]
I can open another PR but it depends on this one.

**Comment by mg979** [comment]
Branch is [here](https://github.com/mg979/neogen/tree/todo_placeholders).

**Comment by danymat** [comment]
I propose you merge into this PR so we ship everything as is when it's done

**Comment by danymat** [comment]
No problem, I will try to work on it today

**Comment by danymat** [comment]
I tried this feature, and it seems to work on snippet. However it's failing on native generator. I think it's best to activate support for this feature in snippets only

**Comment by danymat** [comment]
Seeing this:

image

Makes me want to use `vim.highlight.range` in order to set highlights for placeholders. What do you think ?

**Comment by mg979** [comment]
I really think this should go to another PR. This PR should stop at https://github.com/danymat/neogen/pull/72/commits/033c963bf6d1163389577af39d421c4bc7e52a54

**Comment by danymat** [comment]
Right, I let you revert then

**Comment by mg979** [comment]
I'll pull the current state into the other branch, then I'll force-push at https://github.com/danymat/neogen/commit/033c963bf6d1163389577af39d421c4bc7e52a54

**Comment by mg979** [comment]
https://github.com/danymat/neogen/pull/72/commits/0dfa5f75ebed06272e40324c592f8711e4d1e3e6 is still in https://github.com/mg979/neogen/tree/todo_placeholders

**Comment by danymat** [comment]
Can you take a look ? If you don't like it, feel free to explain and revert the commit. If you're okay, I let you do some testing and I will merge it afterwards

**Comment by mg979** [comment]
Sure it's ok, I did just a little change to check `return_snippet` before `snippet_engine`.

**Comment by danymat** [comment]
Merged ! very good job @mg979, thanks @L3MON4D3 and @max397574 for the help too 🥳

**Comment by danymat** [comment]
Hello @mg979 , mind reopening a new PR for placeholders ? I can work on it if you don't have time

**Pull request merged**
